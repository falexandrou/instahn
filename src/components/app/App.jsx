import React from 'react';

import Header from 'components/header/Header';
import StoryList from 'components/story-list/StoryList';
import Menu from 'components/menu/Menu';
import { fetchStories } from 'actions/stories';
import { STORIES_TOP } from 'app-constants';

import './App.scss';

/**
 * The main app container
 */
class App extends React.Component {
  /**
   * @var {Object} the component's state
   */
  state = {
    activeType: STORIES_TOP,
    isMenuOpen: false,
  };

  /**
   * Changes the type of stories we're viewing
   *
   * @param {String} the type we're switching to
   */
  handleTypeChanged(type) {
    this.setState({
      activeType: type,
      isMenuOpen: false,
    });
  }

  /**
   * Toggles the off canvas menu
   */
  handleMenuToggled() {
    this.setState({ isMenuOpen: !this.state.isMenuOpen });
  }

  render() {
    const { activeType, isMenuOpen } = this.state;

    return <div className="app-container">
      <Menu isOpen={isMenuOpen} onSelectType={ this.handleTypeChanged.bind(this) } activeType={ activeType } />
      <div className="main-wrapper">
        <Header isMenuOpen={ isMenuOpen } onMenuToggled={ this.handleMenuToggled.bind(this) } />
        <div className="content container">
          <StoryList key={`story-list-${activeType}`} type={activeType} />
        </div>
      </div>
    </div>;
  }
}

export default App;