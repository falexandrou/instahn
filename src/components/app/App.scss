@import "./variables.scss";
@import "~bulma/bulma.sass";

body {
  min-height: 101vh; /* force scrollbar to always be visible */
}

.main-wrapper {
  min-width: 100%;
  min-height: 100%;
  position: relative;
  top: 0;
  bottom: 100%;
  left: auto;
  right: 0;
  z-index: 1;
}

@media (max-width: $desktop) {
  .main-wrapper .content {
    margin-left: 1em;
    margin-right: 1em;
  }
}