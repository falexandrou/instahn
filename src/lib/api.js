import {
  STORIES_SHOW,
  STORIES_BEST,
  STORIES_TOP,
  STORIES_NEW,
  STORIES_ASK,
  BASE_API_URL,
} from 'app-constants';

import Cache from './cache';

/**
 * Handles all requests with the external HackerNews API
 * @see https://github.com/HackerNews/API
 * @note All requests are cached
 */
class Api {
  /**
   * @var {Object}
   */
  static uriMapping = {
    [STORIES_SHOW]: `/showstories`,
    [STORIES_BEST]: `/beststories`,
    [STORIES_TOP]: `/topstories`,
    [STORIES_NEW]: `/newstories`,
    [STORIES_ASK]: `/askstories`,
  };

  constructor( cache = null ) {
    this.cache = cache || new Cache();
  }

  /**
   * @param {String} the URI to get the full url for
   * @returns {String} the URL on the HackerNews API
   */
  getUrl(uri) {
    return `${BASE_API_URL}${uri}.json`;
  }

  /**
   * @async
   * @param {String} url
   * @param {Object} data
   * @param {String} method
   * @return {Promise<Response>}
   */
  async request (url) {
    return await fetch(url).then(response => this.cache.set(url, response)).catch(() => this.cache.get(url));
  }

  /**
   * @async
   * @param {String} type of stories to fetch
   * @return {Promise<Response>}
   */
  async fetchStories (type = STORIES_TOP) {
    const uri       = Api.uriMapping[type] || Api.uriMapping[STORIES_TOP];
    const url       = this.getUrl(uri);
    const response  = await this.request(url);

    if (!response)
      return null;

    const list = await response.json();

    if (!list)
      return null;

    return list.sort().reverse();
  }

  /**
   * @async
   * @param {Number} id of a single story to fetch
   * @return {Promise<Response>}
   */
  async fetchStory(id) {
    const url     = this.getUrl(`/item/${id}`);
    let response  = await this.cache.get(url);

    if (!response)
      response = await this.request(url);

    if (!response)
      return null;

    const story = await response.json();

    if (!story || story.type !== 'story')
      return null;

    return story;
  }
}

export default Api;