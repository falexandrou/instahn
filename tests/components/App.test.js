import 'raf/polyfill';
import 'isomorphic-fetch';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from 'components/app/App';
import fetchMock from 'fetch-mock';
import { Provider } from 'react-redux';

import api from '../mocks/api';
import '../mocks/dates';
import { createStore } from '../mocks/store';
import { storyMock } from '../mocks/fixtures';

const fakeStore = createStore({});

describe('App Component', () => {
  beforeEach(() => {
    fetchMock.restore();

    let storiesUrl  = api.getUrl('/topstories'),
        itemUrl     = api.getUrl(`/item/${storyMock.id}`);

    fetchMock.getOnce(storiesUrl, [storyMock.id]);
    fetchMock.getOnce(itemUrl, storyMock);
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Provider store={fakeStore}><App /></Provider>, div);
  });

  it('should match snapshot', () => {
    const output = renderer.create(<Provider store={fakeStore}><App /></Provider>).toJSON();
    expect(output).toMatchSnapshot();
  });
});