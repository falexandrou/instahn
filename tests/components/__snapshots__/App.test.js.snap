// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should match snapshot 1`] = `
<div
  className="app-container"
>
  <nav
    aria-label="main navigation"
    className="navbar is-primary is-transparent"
    role="navigation"
  >
    <div
      className="navbar-brand"
    >
      <a
        className="navbar-item hackernews-logo"
        href="/"
      >
        <img
          alt="Hackernews"
          src="assets/hackernews.svg"
        />
      </a>
    </div>
    <div
      className="navbar-menu"
    >
      <div
        className="navbar-end"
      >
        <a
          className="navbar-item"
          href="https://github.com/falexandrou/instahn"
          target="_blank"
        >
          <img
            alt="GitHub repository"
            src="assets/github.svg"
          />
        </a>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <div
      className="columns"
    >
      <div
        className="column is-three-quarters"
      >
        <div />
      </div>
      <div
        className="column is-one-quarter"
      >
        <aside
          className="menu type-selection"
        >
          <ul
            className="menu-list"
          >
            <li>
              <a
                className="is-active"
                onClick={[Function]}
              >
                Top Stories
              </a>
            </li>
            <li>
              <a
                className=""
                onClick={[Function]}
              >
                New on HackerNews
              </a>
            </li>
            <li>
              <a
                className=""
                onClick={[Function]}
              >
                Best Of
              </a>
            </li>
            <li>
              <a
                className=""
                onClick={[Function]}
              >
                Show HackerNews
              </a>
            </li>
            <li>
              <a
                className=""
                onClick={[Function]}
              >
                Ask HackerNews
              </a>
            </li>
          </ul>
        </aside>
        <div
          className="tags has-addons"
        >
          <span
            className="tag"
          >
            Connectivity
          </span>
          <span
            className="tag is-primary"
          >
            Online!
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
